### ISSUE TITLE
Speed up search by using prefix matching

### ISSUE THREAD

**Issue by jsha** [rust-lang/rust#91076]
Right now, rustdoc search is O(n) in the number of items in the crate, since it runs the matcher across all items:

https://github.com/rust-lang/rust/blob/50f2c292007f9364908e4b8344886797f0144648/src/librustdoc/html/static/js/search.js#L740

That means it's fairly slow, even for the stdlib. For even larger crates, it's even slower. And the slowness is particularly noticeable because it's JavaScript doing compute the whole time, which can cause interactions with the page to freeze. As a result, we have a 500ms timeout on the user typing before we do any searches.

On simple way to make search faster is to use a [trie](https://en.wikipedia.org/wiki/Trie), so if someone types "st" we can quickly find all items that start with that. So for instance the structure in JS might look something like:

```
var trie = {
  "a": {
    "b": {
      "ort": ["process::abort", "intrinsics::abort"],
    },
    "d": {
      "d": {
        "": ["ops::Add"],
        "a": {
          ...
        }
      }
    }
}
```

This could speed up most searches dramatically. It would also allow us to make search match on prefixes. For instance, [right now if I search "add"](https://doc.rust-lang.org/nightly/std/all.html?search=add), I get:

```
std::ops::Add
std::ops::Add::add
pointer::add
std::time::Instant::add
std::time::SystemTime::add
f32::add
f64::add
i8::add
i16::add
i32::add
i64::add
i128::add
u8::add
u16::add
u32::add
u64::add
u128::add
isize::add
usize::add
std::borrow::Cow::add
std::num::Saturating::add
std::num::Wrapping::add
std::simd::Simd::add
```

But I don't get `ops::AddAssign` or `ptr::addr_of`, even though the length of the search results is not maxed out. I think both of those should show up.

The speed improvements could allow us to lower the 500ms typing timeout, making search feel responsive; or we could keep it, and simply enjoy better performance for huge crates.

Note: If we want to get the speed improvements of searching by trie, that means either (a) giving up levenshtein matching, or (b) doing levenshtein matching in a second phase, if the results from prefix matching weren't enough to fill the page.

/cc @GuillaumeGomez

**Comment by GuillaumeGomez** [comment]
That would still be `O(n)` but the different checks run inside would be different. An important thing to note is that the current rustdoc search already look if the search string is in the item.

**Comment by jsha** [comment]
It would be O(m) where `m` is the number of results (or characters in the search term, but there will generally be more results than characters in the search term). Here's a simplified example of how looking up results for "add" would go (emitting error checking):

```
var results = trie['a']['d']['d'];
```

Note that it would still be O(n) to build the tree (or to parse it, if we built it on the Rust side and sent it down as JSON). But the JSON parser is quite fast. And having search be faster even if parsing is slow is useful because we can provide better interactivity and search-as-you-type.

**Comment by GuillaumeGomez** [comment]
I'm not sure it's worth it because it forces the search to behave in a given way. Maybe we should try to discuss how the search should behave globally and then once we have something, try to improve performance. In here, it seems like we're doing things in the wrong order.

**Comment by jsha** [comment]
This is a good point - as the saying goes, post problems rather than solutions. So, in that spirit:

Problem: rustdoc search is too slow, particularly when searching many crates.

I went to find a citation for that - for instance, I though I saw someone mention a use case where search was very slow for them. However, I couldn't find it! I thought serde might be one such example but https://docs.serde.rs/serde/index.html?search=json renders very fast!

So I'll close this for now; speed (other than download speed) seems to be a non-issue.
